src/hardware/signal_generator.rs

```diff
@@ -0,0 +1,176 @@
+use crate::hardware::{dac::DacCode, design_parameters::ADC_SAMPLE_TICKS};
+use core::convert::{TryFrom, TryInto};
+use miniconf::Miniconf;
+use serde::Deserialize;
+
+/// Types of signals that can be generated.
+#[derive(Copy, Clone, Debug, Deserialize, Miniconf)]
+pub enum Signal {
+    Cosine,
+    Square,
+    Triangle,
+}
+
+/// Basic configuration for a generated signal.
+///
+/// # Miniconf
+/// `{"signal": <signal>, "frequency", 1000.0, "symmetry": 0.5, "amplitude": 1.0}`
+///
+/// Where `<signal>` may be any of [Signal] variants, `frequency` specifies the signal frequency
+/// in Hertz, `symmetry` specifies the normalized signal symmetry which ranges from 0 - 1.0, and
+/// `amplitude` specifies the signal amplitude in Volts.
+#[derive(Copy, Clone, Debug, Miniconf, Deserialize)]
+pub struct BasicConfig {
+    /// The signal type that should be generated. See [Signal] variants.
+    pub signal: Signal,
+
+    /// The frequency of the generated signal in Hertz.
+    pub frequency: f32,
+
+    /// The normalized symmetry of the signal. At 0% symmetry, the first half phase does not exist.
+    /// At 25% symmetry, the first half-phase lasts for 25% of the signal period.
+    pub symmetry: f32,
+
+    /// The amplitude of the output signal in volts.
+    pub amplitude: f32,
+}
+
+impl Default for BasicConfig {
+    fn default() -> Self {
+        Self {
+            frequency: 1.0e3,
+            symmetry: 0.5,
+            signal: Signal::Cosine,
+            amplitude: 0.0,
+        }
+    }
+}
+
+/// Represents the errors that can occur when attempting to configure the signal generator.
+#[derive(Copy, Clone, Debug)]
+pub enum Error {
+    /// The provided amplitude is out-of-range.
+    InvalidAmplitude,
+}
+
+impl TryFrom<BasicConfig> for Config {
+    type Error = Error;
+
+    fn try_from(config: BasicConfig) -> Result<Config, Error> {
+        // Calculate the frequency tuning words
+        let frequency_tuning_word: [u32; 2] = {
+            let conversion_factor =
+                ADC_SAMPLE_TICKS as f32 / 100.0e6 * (1u64 << 32) as f32;
+
+            if config.symmetry <= 0.0 {
+                [
+                    i32::MAX as u32,
+                    (config.frequency * conversion_factor) as u32,
+                ]
+            } else if config.symmetry >= 1.0 {
+                [
+                    (config.frequency * conversion_factor) as u32,
+                    i32::MAX as u32,
+                ]
+            } else {
+                [
+                    (config.frequency * conversion_factor / config.symmetry)
+                        as u32,
+                    (config.frequency * conversion_factor
+                        / (1.0 - config.symmetry)) as u32,
+                ]
+            }
```

Comment:
Streamlining the math a bit:

```suggestion
            const LSB_PER_HERTZ: f32 = ((1u64 + ADC_SAMPLE_TICKS_LOG2) << 32) as f32 / 100.0e6;
            let ftw = config.frequency * LSB_PER_HERTZ;

            if config.symmetry <= 0.0 {
                [1u32 << 31, ftw as u32]
            } else if config.symmetry >= 1.0 {
                [ftw as u32, 1u32 << 31]
            } else {
                [
                    (ftw / config.symmetry) as u32,
                    (ftw / (1.0 - config.symmetry)) as u32,
                ]
            }
```